File: public/codes/design-pattern/creational/prototype.py
import copy

class Prototype:
    def __init__(self) -> None:
        self.primitive = None
        self.component = None
        self.circular_reference = None

    def __copy__(self) -> object:
        # Create a shallow copy of the object.
        new = self.__class__()
        new.__dict__.update(self.__dict__)
        return new

    def __deepcopy__(self, memo: dict = {}) -> object:
        # Create a deep copy of the object.
        new = self.__class__()
        memo[id(self)] = new
        for k, v in self.__dict__.items():
            setattr(new, k, copy.deepcopy(v, memo))
        return new

class ComponentWithBackReference:
    def __init__(self, prototype: Prototype) -> None:
        self.prototype = prototype

File: public/codes/design-pattern/creational/prototype.ts
class Prototype {
  public primitive: any;
  public component: object;
  public circularReference: ComponentWithBackReference;

  constructor(primitive: any, component: object, circularReference: ComponentWithBackReference) {
    this.primitive = primitive;
    this.component = component;
    this.circularReference = circularReference;
  }

  public clone(): this {
    const clone = Object.create(this);

    clone.component = Object.create(this.component);

    // Clone an object that has a circular reference. Meticulously clone the
    // circular reference object, otherwise the cloned object will point to the
    // original object's circular reference object.
    clone.circularReference = Object.create(this.circularReference);
    clone.circularReference.prototype = {
      ...Object.create(this.circularReference.prototype),
      circularReference: clone,
    };

    return clone;
  }
}

class ComponentWithBackReference {
  public prototype: Prototype;

  constructor(prototype: Prototype) {
    this.prototype = prototype;
  }
}
